asset_dashboard/templates/asset_dashboard/partials/forms/add_edit_phase_form.html

```diff
@@ -129,6 +154,239 @@ <h3>Phase Assets</h3>
+              </td>
+              <td class='pl-2'>
+                {{ form.funding_streams.funding_secured }}
+                <div class="secured-error" aria-hidden="true"></div>
+              </td>
+              <td>
+                <a href='#/' class='btn btn-sm btn-outline-danger' onclick='removeForm(this);'><i class='fas fa-times' aria-hidden="true"></i> Remove</a>
+              </td>
+            </tbody>
+          </table>
+        </form>`
+
+      // Create a new form node from a string
+      const range = document.createRange()
+      const fragment = range.createContextualFragment(newForm)
+      let newTr = document.createElement("tr")
+      let newTd = document.createElement("td")
+      newTr.classList.add('new-tr')
+      newTd.setAttribute('colspan', '100%')
+
+      // Prevent previous unsaved inputs from being cleared when adding a new form
+      newTd.appendChild(fragment)
+      newTr.appendChild(newTd)
+      fundingTable.appendChild(newTr)
+    }
+
+    function removeForm(target) {
+      target.closest('.new-tr').remove()
+    }
+
+    function submitFundingForms() {
+      fundings = document.getElementsByClassName("funding")
+      serverError = document.getElementById("server-error")
+      let valid = true
+
+      // Front end validation for each form on the page
+      for (i = 0; i < fundings.length; i++) {
+        let funding = fundings[i]
+        let year = funding[0].value
+        let budget = funding[1].value
+        let actual_cost = funding[3].value
+
+        const yearError = funding[0].nextElementSibling
+        const budgetError = funding[1].nextElementSibling.nextElementSibling
+        const costError = funding[3].nextElementSibling.nextElementSibling
+        const sourceError = funding[5].nextElementSibling
+        const securedError = funding[6].nextElementSibling
+
+        // Hide the errors that front end validation doesn't consider, 
+        // in case they're visible from a previously failed attempt
+        hideError(sourceError)
+        hideError(securedError)
+        
+        if (year.toString().length != 4 || !year){
+          valid = false
+          yearError.innerHTML = 'Please enter a valid year'
+          showError(yearError)
+        } else {
+          hideError(yearError)
+        }
+
+        if (budget <= 0){
+          valid = false
+          budgetError.innerHTML = 'Please enter a valid amount'
+          showError(budgetError)
+        } else {
+          hideError(budgetError)
+        }
+
+        if (actual_cost <= 0){
+          valid = false
+          costError.innerHTML = 'Please enter a valid amount'
+          showError(costError)
+        } else {
+          hideError(costError)
+        }
+      }
+
+      // Back end validation and submission of forms
+      if (valid) {
+        let backendValid = true
+        let fetches = []
+
+        for (i = 0; i < fundings.length; i++) {
```

Comment:
did you try a .forEach or .map here? my hunch is that the async nature of fetch might not work well with those array functions (i've had weird problems trying to do async stuff within forEach).

just curious if you tried, no change necessary.

Reply:
it looks like since i'm grabbing this with `getElementsByClassName` it's returning an htmlcollection, and i remember trying `foreach` with an htmlcollection before (not sure if here or elsewhere) and it not working.

turns out those don't have `foreach` or `map` methods, [as opposed to a nodelist](https://www.w3schools.com/js/js_htmldom_nodelist.asp#:~:text=The%20querySelectorAll()%20method%20returns,property%20returns%20a%20live%20NodeList.) which i could've gotten if i used `querySelectorAll` instead. i never knew this was the reason, TIL